pre_commit/languages/rust.py

```diff
@@ -16,27 +22,62 @@
 from pre_commit.prefix import Prefix
 from pre_commit.util import clean_path_on_failure
 from pre_commit.util import cmd_output_b
+from pre_commit.util import make_executable
+from pre_commit.util import win_exe
 
 ENVIRONMENT_DIR = 'rustenv'
-get_default_version = helpers.basic_get_default_version
-health_check = helpers.basic_health_check
+TOOLCHAIN_FROM_VERSION = {
+    'system': None,
+    C.DEFAULT: 'stable',
+}
 
 
-def get_env_patch(target_dir: str) -> PatchesT:
+@functools.lru_cache(maxsize=1)
+def get_default_version() -> str:
+    # if ust is already installed, we can save a bunch of setup time by
+    # using the installed version
+    if helpers.exe_exists('cargo'):
+        return 'system'
+    else:
+        return C.DEFAULT
+
+
+def _envdir(prefix: Prefix, version: str) -> str:
+    directory = helpers.environment_dir(ENVIRONMENT_DIR, version)
+    return prefix.path(directory)
+
+
+def get_env_patch(target_dir: str, version: str) -> PatchesT:
+    toolchain = TOOLCHAIN_FROM_VERSION.get(version, version)
     return (
+        ('CARGO_HOME', target_dir),
         ('PATH', (os.path.join(target_dir, 'bin'), os.pathsep, Var('PATH'))),
+        # Only set RUSTUP_TOOLCHAIN if we don't want use the system's default
+        # toolchain
+        *((('RUSTUP_TOOLCHAIN', toolchain),) if toolchain else ()),
     )
 
 
 @contextlib.contextmanager
-def in_env(prefix: Prefix) -> Generator[None, None, None]:
-    target_dir = prefix.path(
-        helpers.environment_dir(ENVIRONMENT_DIR, C.DEFAULT),
-    )
-    with envcontext(get_env_patch(target_dir)):
+def in_env(
+    prefix: Prefix,
+    language_version: str,
+) -> Generator[None, None, None]:
+    with envcontext(
+        get_env_patch(_envdir(prefix, language_version), language_version),
+    ):
         yield
 
 
+def health_check(prefix: Prefix, language_version: str) -> str | None:
+    with in_env(prefix, language_version):
+        retcode, _, _ = cmd_output_b('cargo', '--version', retcode=None)
+        if retcode != 0:
+            return f'`cargo --version` returned {retcode}'
+        else:
+            return None
```

Comment:
`health_check` might not be needed, though curious to hear your thoughts (most of the `health_check` implementations are for situations where OS upgrades pull out the necessary parts of built environments) -- think like `libssl1.1` -> `libssl3.0` and then the `.so` files are missing links.  my understanding of rust is that it's self-contained ?

Reply:
https://github.com/pre-commit/pre-commit/pull/2534/commits/901e7541dc2383ee075babb8d6d069fb8f73e477